NOAA-EMC/regional_workflow
# Fix directory overwrite bug in NCO mode.

This needs to be merged after PR #221.

Summary of modifications:
------------------------
* Modify scripts to prevent simultaneously running experiments in NCO mode from writing to the same set of directories.

Tests conducted:
---------------
* All WE2E tests pass except regional_013. regional_013 doesn't pass because the physics suite used by that test (FV3_GSD_SAR_v1) doesn't actually exist in the ufs-weather-model repo.

Modifications common to more than one file (used below in listing of file-by-file modifications):
------------------------------------------------------------------------------------------------
(A) Fix/add/delete comments and/or informational and/or error messages.
(B) Remove unused code and/or fix code indentation.
(C) Set the RUN and envir variables to the name of the WE2E test name contained in EXPT_SUBDIR.  Note that even though EXPT_SUBDIR is set to a null string near the beginning of this file, the script that runs the WE2E tests inserts the name of the test into the definition of EXPT_SUBDIR (i.e. it changes the line 'EXPT_SUBDIR=""' into 'EXPT_SUBDIR="test_name"') before calling the experiment generation script.  Thus, EXPT_SUBDIR will be properly set to the name of the test when the experiment generation script is called.
(D) Bug fix - change variable name from LBC_UPDATE_INTVL_HRS to LBC_SPEC_INTVL_HRS.


File-by-file description of modifications:
-----------------------------------------

jobs/JREGIONAL_RUN_POST:
* In the definition of COMOUT, use the new workflow variable COMOUT_BASEDIR that specifies the base directory under which each cycle's UPP output directories will be placed.

tests/baseline_configs/config.nco_conus.sh: (A), (C)

tests/baseline_configs/config.nco_conus_c96.sh: (A), (C)

tests/baseline_configs/config.regional_009.sh: (A), (C)
* Change the time interval for specifying external model LBCs from 6 hrs to 3 hrs (to match what is done in all the other WE2E tests that use FV3GFS as the external model for LBCs).

tests/run_experiments.sh: (A), (B)

ush/config.community.sh: (D)

ush/config.nco.sh: (D)

ush/config_defaults.sh: (A)

ush/setup.sh: (A)
* For simplicity, combine two blocks of code that create different directories into a single block (that creates both sets of directories).
* In NCO mode, change defintion of CYCLE_BASEDIR from "$STMP/tmpnwprd/${EMC_GRID_NAME}" to "$STMP/tmpnwprd/$RUN".  This is to make it easier to associate this directory with an experiment (wich can be done by setting RUN to the experiment name in the experiment's configuration file).  That way, two experiments running simultaneously (but having different names) will not end up writing to the same CYCLE_BASEDIR (which currently can happen if the two experiments are using the same grid, i.e. the same value of EMC_GRID_NAME).
* Add COMOUT_BASEDIR as a new secondary workflow variable.  In NCO mode, set this to the base directory under which the output from the RUN_POST_TN task will be placed (i.e. the cycle-independent portion of the RUN_POST_TN task's output directory) and then check whether this directory exists (and deal with it as specified by the variable PREEXISTING_DIR_METHOD).  The goal here is to be able to uniquely associate this directory to an experiment so that simultaneously running experiments do not write their UPP outputs into the same directory (and possibly overwrite each other's results).  This directory is given by COMOUT_BASEDIR="$COMROOT/$NET/$envir", so we can uniquely associate it with an experiment by setting envir to the name of the experiment (which can be done in the experiment's configuration file).  This is why in the WE2E configuration files listed above (config.nco_conus.sh, config.nco_conus_c96.sh, and config.regional_009.sh), we set envir to the name of the experiment.  In community mode, the new variable COMOUT_BASEDIR is not used, so in that case we set it to an empty string.
